### PR TITLE
Add tc_reply_kern_02 program to adjust reply packets' source port

### PR DESCRIPTION
**Pull request by bigclouds**
As xdp_prog_kern_02.c modifies destination port , this tc program is to fix reply packets.
It needs to work together with xdp_prog_kern_02.

### CODE REVIEW

**Comment by bigclouds** [comment]
@tohojo Hi, please take time to review this PR. Thanks

File: packet-solutions/tc_reply_kern_02.c [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+{
+	struct iphdr *iphdr;
+	struct ipv6hdr *ipv6hdr;
+        struct udphdr *udphdr;
+        struct tcphdr *tcphdr;
+	int eth_type, ip_type;
+	void *data = (void *)(long)skb->data;
+	struct ethhdr *eth;
+	void *data_end = (void *)(long)skb->data_end;
+	struct hdr_cursor nh = { .pos = data };
```
Comment (tohojo) [inline packet-solutions/tc_reply_kern_02.c:24]: We follow the netdev "reverse christmas tree" variable ordering, see https://docs.kernel.org/process/maintainer-netdev.html#local-variable-ordering-reverse-xmas-tree-rcs

Please reorder these accordingly...

File: packet-solutions/tc_reply_kern_02.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <linux/if_ether.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+#include <linux/pkt_cls.h>
+
+// The parsing helper functions from the packet01 lesson have moved here
+#include "../common/parsing_helpers.h"
+
+SEC("fix_port_egress")
```
Comment (tohojo) [inline packet-solutions/tc_reply_kern_02.c:13]: Custom section names have been deprecated in libbpf. This should be `SEC("tc")`

File: packet-solutions/tc_reply_kern_02.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+	int ret = TC_ACT_OK;
+
+	if (data + sizeof(*eth) > data_end)
+                goto out;
+
+	eth_type = parse_ethhdr(&nh, data_end, &eth);
+        if (eth_type < 0) {
+                goto out;
+        }
```
Comment (tohojo) [inline packet-solutions/tc_reply_kern_02.c:34]: No need for the braces

File: packet-solutions/tc_reply_kern_02.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	} else if (eth_type == bpf_htons(ETH_P_IPV6)) {
+		ip_type = parse_ip6hdr(&nh, data_end, &ipv6hdr);
+	} else {
+		goto out;
+	}
+
+	if (ip_type == IPPROTO_UDP) {
+		if (parse_udphdr(&nh, data_end, &udphdr) < 0) {
+			goto out;
+		}
```
Comment (tohojo) [inline packet-solutions/tc_reply_kern_02.c:47]: Unnecessary braces

File: packet-solutions/tc_reply_kern_02.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		udphdr->source = bpf_htons(bpf_ntohs(udphdr->source) + 1);
+		udphdr->check += bpf_htons(-1);
+		if (!udphdr->check)
+			udphdr->check += bpf_htons(-1);
+	} else if (ip_type == IPPROTO_TCP) {
+		if (parse_tcphdr(&nh, data_end, &tcphdr) < 0) {
+			goto out;
+		}
```
Comment (tohojo) [inline packet-solutions/tc_reply_kern_02.c:55]: Unnecessary braces

**Review by tohojo — CHANGES_REQUESTED** [review]
Some nits, mostly. How did you test this?

**Comment by bigclouds** [comment]
@tohojo Thanks. All Done.

**Pull request merged**